#' @importFrom Rcpp evalCpp
#' @importFrom Matrix colSums rowSums colMeans rowMeans
#' @importFrom methods setClass setOldClass setClassUnion slot
#' slot<- setMethod new signature slotNames is
#' @importClassesFrom Matrix dgCMatrix


#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# Class definitions
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

setOldClass(Classes = 'package_version')
setClassUnion(name = 'AnyMatrix', c("matrix", "dgCMatrix"))


### Internal functions
# Left hand side or right hand side
#
# @param lhs left hand side
# @param rhs right hand side
#
`%||%` <- function(lhs, rhs) {
  if (!is.null(x = lhs)) {
    return(lhs)
  } else {
    return(rhs)
  }
}


#' The Seurat Version 2 Class
#'
#' The Seurat object is a representation of single-cell expression data for R; each Seurat
#' object revolves around a set of cells and consists of one or more objects, or individual 
#' representations of expression data (eg. RNA-seq, ATAC-seq, etc).
#' These assays can be reduced from their high-dimensional state to a lower-dimension state
#' and stored as \code{\link{DimReduc-class}} objects. Seurat objects also store additional
#' meta data, both at the cell and feature level (contained within individual assays). The
#' object was designed to be as self-contained as possible, and easily extendible to new methods.
#'
#' @slot raw.data A list of assays for this project
#' @slot data
#' @slot scale.data
#' @slot var.genes
#' @slot is.expr
#' @slot meta.data Contains meta-information about each cell, starting with number of genes detected (nGene)
#' and the original identity class (orig.ident)
#' @slot ident The active cluster identity for this Seurat object
#' @slot dr A list of dimmensional reduction objects for this object
#' @slot assay A list of assays
#' @slot hvg.info
#' @slot project.name Name of the project
#' @slot misc A list of miscellaneous information
#' @slot version Version of Seurat this object was built under
#' @slot commands A list of logged commands run on this \code{Seurat} object
#' @slot tools A list of miscellaneous data generated by other tools, should be filled by developers only using 
#'
#' @name Seurat-class-v2
#' @rdname Seurat-class
#' @exportClass seurat
#'
seurat <- methods::setClass(
  "seurat",
  slots = c(
    raw.data = "ANY",
    data = "ANY",
    scale.data = "ANY",
    var.genes = "vector",
    is.expr = "numeric",
    ident = "factor",
    meta.data = "data.frame",
    project.name = "character",
    dr = "list",
    assay = "list",
    hvg.info = "data.frame",
    imputed = "data.frame",
    cell.names = "vector",
    cluster.tree = "list",
    snn = "dgCMatrix",
    calc.params = "list",
    kmeans = "ANY",
    spatial = "ANY",
    misc = "ANY",
    version = "ANY"
  )
)

#' The Dimmensional Reduction Class
#'
#' The DimReduc object stores a dimensionality reduction taken out in Seurat; each DimReduc
#' consists of a cell embeddings matrix, a feature loadings matrix, and a projected feature
#' loadings matrix.
#'
#' @slot cell.embeddings Cell embeddings matrix (required)
#' @slot gene.loadings Feature loadings matrix (optional)
#' @slot gene.loadings.full 
#' @slot sdev A vector of standard deviations
#' @slot gene.loadings.full
#' @slot key Key for the DimReduc, must be alphanumerics followed by an underscore
#' @slot jackstraw A \code{\link{JackStrawData-class}} object associated with this DimReduc
#' @slot misc Utility slot for storing additional data associated with the DimReduc
#'       (e.g. the total variance of the PCA)
#'
#' @name DimReduc-class
#' @rdname DimReduc-class
#' @exportClass dim.reduction
#'
dim.reduction <- setClass(
  Class = "dim.reduction",
  slots = list(
    cell.embeddings = "matrix",
    gene.loadings = "matrix",
    gene.loadings.full = "matrix",
    sdev = "numeric",
    key = "character",
    jackstraw="ANY",
    misc = "ANY"
  )
)



#' The jackstraw.data class
#'
#' The JackStrawData is used to store the results of a JackStraw computation.
#'
#' @slot empirical.p.value Empirical p-values
#' @slot fake.pc.scores Fake reduction scores
#' @slot empirical.p.value.full Empirical p-values on full
#' 
#'
#' @name jackstraw.data-class
#' @rdname jackstraw.data-class
#' @exportClass jackstraw.data
#'

jackstraw.data <- setClass(
  Class = "jackstraw.data",
  slots = list(
    emperical.p.value = "matrix",
    fake.pc.scores = "matrix",
    emperical.p.value.full = "matrix"
  )
)


# DowngradeDimReduction
#
# @param old.dr 
DowngradeDimReduction <- function(old.dr){
  new.dr <- list()
  keynames = list(pca = "PC", tsne = "tSNE_", umap = "UMAP_")
  for (i in names(x = old.dr)) {
    cell.embeddings <- old.dr[[i]]@cell.embeddings %||% new(Class = 'matrix')
    gene.loadings <- old.dr[[i]]@feature.loadings %||% new(Class = 'matrix')
    sdev <- old.dr[[i]]@stdev %||% numeric()
    misc <- old.dr[[i]]@misc %||% list()
    new.jackstraw <- SCFunctionsV3:::DowngradeJackstraw(old.jackstraw = old.dr[[i]]@jackstraw)
    old.key <- old.dr[[i]]@key
    new.key <- keynames[[i]]
    colnames(x = cell.embeddings) <- gsub(
      pattern = old.key,
      replacement = new.key,
      x = colnames(x = cell.embeddings)
    )
    colnames(x = gene.loadings) <- gsub(
      pattern = old.key,
      replacement = new.key,
      x = colnames(x = gene.loadings)
    )
    new.dr[[i]] <- new(
      Class = 'dim.reduction',
      cell.embeddings = as(object = cell.embeddings, Class = 'matrix'),
      gene.loadings = as(object = gene.loadings, Class = 'matrix'),
      sdev = as(object = sdev, Class = 'numeric'),
      key = as(object = new.key, Class = 'character'),
      jackstraw = new.jackstraw,
      misc = as(object = misc, Class = 'list')
    )
  }
  return(new.dr)
}

# DowngradeJackstraw
#
# @param old.jackstraw 
DowngradeJackstraw <- function(old.jackstraw) {
  if (is.null(x = old.jackstraw)) {
    new.jackstraw <- new(
      Class = 'jackstraw.data',
      fake.pc.scores = new(Class = 'matrix'),
      emperical.p.value = new(Class = 'matrix'),
      emperical.p.value.full = new(Class = 'matrix')
    )
  } else {
    new.jackstraw <- new(
      Class = 'jackstraw.data',
      emperical.p.value = old.jackstraw@empirical.p.values %||% new(Class = 'matrix'),
      fake.pc.scores = old.jackstraw@fake.reduction.scores %||% new(Class = 'matrix'),
      emperical.p.value.full = old.jackstraw@empirical.p.values.full %||% new(Class = 'matrix')
    )
  }
  return(new.jackstraw)
}



